File: src/main/java/io/github/chindeaone/collectiontracker/commands/SkillTracker.java
package io.github.chindeaone.collectiontracker.commands;

import io.github.chindeaone.collectiontracker.api.hypixelapi.SkillApiFetcher;
import io.github.chindeaone.collectiontracker.util.SkillUtils;
import io.github.chindeaone.collectiontracker.tracker.skills.SkillTrackingHandler;
import io.github.chindeaone.collectiontracker.util.ChatUtils;
import io.github.chindeaone.collectiontracker.util.HypixelUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.concurrent.CompletableFuture;

import static io.github.chindeaone.collectiontracker.tracker.skills.SkillTrackingHandler.isPaused;
import static io.github.chindeaone.collectiontracker.tracker.skills.SkillTrackingHandler.isTracking;

public class SkillTracker {

    public static String skillName = "";

    public static Logger logger = LogManager.getLogger(SkillTracker.class);

    public static void startTracking(String skill) {
        try {
            if (!HypixelUtils.isOnSkyblock()) {
                ChatUtils.INSTANCE.sendMessage("§cYou must be on Hypixel Skyblock to use this command!", true);
                return;
            }

            try {
                if (isTracking || isPaused) {
                    ChatUtils.INSTANCE.sendMessage("§cAlready tracking a skill.", true);
                    return;
                }

                skillName = skill;
                if (!SkillUtils.isValidSkill(skillName)) {
                    ChatUtils.INSTANCE.sendMessage("§4" + skillName + " skill is not a real skill!", true);
                    return;
                }

                // Fetch skill data asynchronously and start tracking once done
                CompletableFuture.runAsync(SkillApiFetcher::fetchSkillsData)
                        .thenRun(SkillTrackingHandler::startTracking);
            } catch (Exception e) {
                logger.error("An error occurred while starting skill tracking: ", e);
                ChatUtils.INSTANCE.sendMessage("§cAn error occurred while starting skill tracking. Please try again later.", true);
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/java/io/github/chindeaone/collectiontracker/commands/SkillTracker.kt
package io.github.chindeaone.collectiontracker.commands

import io.github.chindeaone.collectiontracker.api.hypixelapi.SkillApiFetcher
import io.github.chindeaone.collectiontracker.config.ConfigAccess
import io.github.chindeaone.collectiontracker.tracker.skills.SkillFetcher
import io.github.chindeaone.collectiontracker.utils.SkillUtils
import io.github.chindeaone.collectiontracker.tracker.skills.SkillTrackingHandler
import io.github.chindeaone.collectiontracker.utils.chat.ChatUtils
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import java.util.concurrent.CompletableFuture

import io.github.chindeaone.collectiontracker.tracker.skills.SkillTrackingHandler.isTracking

object SkillTracker {

    var skillName = ""
    private val logger: Logger = LogManager.getLogger(SkillTracker::class.java)

    fun startTracking(skill: String) {
        try {
            if (isTracking) {
                ChatUtils.sendMessage("§cAlready tracking a skill.", true)
                return
            }

            skillName = skill
            if (!SkillUtils.isValidSkill(skillName)) {
                ChatUtils.sendMessage("§4$skillName skill is not a real skill!", true)
                skillName = ""
                return
            }

            // Fetch skill data and leaderboard data asynchronously
            SkillApiFetcher.fetchSkillsData()
                    .thenCompose { SkillFetcher.fetchSkillLeaderboardData(skillName) }
                    .thenCompose {
                        if (ConfigAccess.isTamingTrackingEnabled()) SkillFetcher.fetchSkillLeaderboardData("Taming")
                        else CompletableFuture.completedFuture(null)
                    }
                    .thenRun(SkillTrackingHandler::startTracking)
        } catch (e: Exception) {
            logger.error("An error occurred while starting skill tracking: ", e)
            ChatUtils.sendMessage("§cAn error occurred while starting skill tracking. Please try again later.", true)
        }
    }
}